src/headers.rs

```diff
@@ -64,6 +58,13 @@ pub(super) fn content_length_parse_all_values(values: ValueIter<'_, HeaderValue>
     }
 }
 
+fn content_length_parse_line(value: &HeaderValue) -> Vec<Result<u64, ()>> {
```

Comment:
The creation of a `Vec` here will mean a heap allocation, which will slow things down a bit more. I believe it should be possible to still do this without that. It might end up making the combinators a nightmare though, so it may eventually be worth just sprinkling a change to a `for ...` and an early return instead... If you get stuck, I can see what I can come up with.

Reply:
Hi,

Here is an other version without Vec:
```rust
pub(super) fn content_length_parse_all_values(values: ValueIter<'_, HeaderValue>) -> Option<u64> {
    // If multiple Content-Length headers were sent, everything can still
    // be alright if they all contain the same value, and all parse
    // correctly. If not, then it's an error.

    let fold_values = |prev: Option<Result<u64, ()>>, value: Result<u64, ()>| match prev {
        Some(Ok(prev)) => Some(
            value.and_then(|n| if prev == n { Ok(n) } else { Err(()) }),
        ),
        None => Some(value),
        Some(Err(())) => Some(Err(())),
    };

    let folded = values.fold(None, |prev, line| {
        return match line.to_str() {
            Ok(line) => {
                return line.split(',')
                    .map(|a| a.trim().parse().map_err(|_| ()))
                    .fold(None, fold_values)
                    .map(|v| fold_values(prev, v))
                    .unwrap_or(None);
            },
            Err(_) => Some(Err(()))
        };
    });

    if let Some(Ok(n)) = folded {
        Some(n)
    } else {
        None
    }
}
```
What do you think about this?

Reply:
Nice, that looks like it doesn't need to allocate, awesome!

Looking at this code, it does seem rather complicated trying to keep using `fold`. I'll see if I can come up with a version that uses a `for` loop that's less complicated...

Reply:
Here is a version with nested loops and early returns:
```rust
pub(super) fn content_length_parse_all_values(values: ValueIter<'_, HeaderValue>) -> Option<u64> {
    // If multiple Content-Length headers were sent, everything can still
    // be alright if they all contain the same value, and all parse
    // correctly. If not, then it's an error.

    let mut content_length: Option<u64> = None;
    for h in values {
        if let Ok(line) = h.to_str() {
            for v in line.split(',') {
                if let Some(n) = v.trim().parse().ok() {
                    if content_length.is_none() {
                        content_length = Some(n)
                    } else if content_length != Some(n) {
                        return None;
                    }
                } else {
                    return None
                }
            }
        } else {
            return None
        }
    }

    return content_length
}
```

Reply:
Oh awesome! Yea, this looks so much simpler than the `fold` that was there. And with the extra fold. Thanks, let's use this one!